### ISSUE TITLE
0.8.2 build failure on NetBSD

### ISSUE THREAD

**Issue by 0323pin** [sstadick/hck#64]
Hi,

What I thought would be a quick update has turned into a headache.

For background see, https://github.com/orhun/git-cliff/issues/105 and https://github.com/rust-lang/git2-rs/issues/881

Now linking against `libgit2` and `openssl` doesn't seem to help anymore.
Here's my current package Makefile

```
# $NetBSD: Makefile,v 1.3 2022/06/08 11:33:42 pin Exp $

DISTNAME=	hck-0.8.2
CATEGORIES=	textproc
MASTER_SITES=	${MASTER_SITE_GITHUB:=sstadick/}
GITHUB_TAG=	v${PKGVERSION_NOREV}

MAINTAINER=	pin@NetBSD.org
HOMEPAGE=	https://github.com/sstadick/hck/
COMMENT=	Sharp cut(1) clone
LICENSE=	unlicense OR mit

BUILD_DEPENDS+=		cmake-[0-9]*:../../devel/cmake

.include "cargo-depends.mk"

USE_TOOLS+=		pkg-config

INSTALLATION_DIRS=	bin

MAKE_ENV+=	OPENSSL_DIR=${BUILDLINK_PREFIX.openssl:Q}
RUSTFLAGS+=	-C link-arg=${COMPILER_RPATH_FLAG}${BUILDLINK_PREFIX.openssl}/lib
RUSTFLAGS+=	-C link-arg=${COMPILER_RPATH_FLAG}${BUILDLINK_PREFIX.libgit2}/lib
#RUSTFLAGS+=	-C link-arg=${COMPILER_RPATH_FLAG}${BUILDLINK_PREFIX.libgit2}/include
#RUSTFLAGS+=	-C link-arg=-L${BUILDLINK_PREFIX.libgit2}/include

do-install:
	${INSTALL_PROGRAM} ${WRKSRC}/target/release/hck \
		${DESTDIR}${PREFIX}/bin

.include "../../lang/rust/cargo.mk"
.include "../../devel/libgit2/buildlink3.mk"
.include "../../security/openssl/buildlink3.mk"
.include "../../mk/bsd.pkg.mk"
```

And the error I'm getting, https://pastebin.com/Kg4KzDK3
I know this is not related to `hck` but, maybe you have an idea?

Thanks.

**Comment by 0323pin** [comment]
https://github.com/libgit2/libgit2/issues/6457

**Comment by sstadick** [comment]
Uffff, that's painful. libgit2 has provided nothing but pain and suffering. It's only used to provide a version string that includes the git hash of the commit that the bin was built with. Since creating this tool, I've found a better way with `git-version`, which is more of a soft dep on git the CLI tool and introduces wayyy less build complexity.

I'll try to move this tool to `git-version` in the next day or two and release a `v0.9.0` version that you should skip to.

**Comment by 0323pin** [comment]
@sstadick Thank you 👍

The worst of all, the function is never used see, https://github.com/libgit2/libgit2/pull/6458

As for packages is _hit or miss_, I haven't found a way to tell when it will fail. `hck`-0.7.5 builds fine, while 0.8.2 does not. But, both have a dependency on `libgit2-sys`.

`git-cliff` builds but, `cocogitto` fails. Anyway, ...

> I'll try to move this tool to git-version in the next day or two

Sounds like a good plan.

**Comment by ethomson** [comment]
> Uffff, that's painful. libgit2 has provided nothing but pain and suffering.

I'm sorry to hear that. What other problems have you had? How can we help improve other peoples experience in the future?

**Comment by sstadick** [comment]
@ethomson - I shouldn't have been so harsh - I know you are all doing a lot of work behind the scenes. 

It's areas like this: https://github.com/conda-forge/cargo-bundle-licenses-feedstock/pull/6 that cause me a lot of frustration with libgit2, but that can be generalized to any non-rust dependency really, and a lot of it tracks back to my own ignorance of the low level details with how things are linked / built. A better framing here would be for me to say: that's painful, I don't understand how the dependency on libgit2 actually works.

So in this particular instance, it's easier to just remove libgit2 as a dependency since it's entirely tangential to the purpose of this tool, and adds a lot of complexity in the build process.

How can we improve the experience in the future: I don't know enough to have a good answer here. Maybe @0323pin has some thoughts though since some of this is more fresh-in-mind for them?

**Comment by 0323pin** [comment]
I think the most puzzling and frustrating bit is, as mentioned already, I can't (or, don't know how to) predict if a build will fail or not when there's a dependency on `libgit2-sys`.

Like `hck`-0.7.5 builds when depending on `libgit2-sys-0.12.26+1.3.0` while, the latest 0.8.2. doesn't when depending on `libgit2-sys-0.14.2+1.5.1`.

Also, as already mentioned, I can  build `git-cliff`, see https://github.com/orhun/git-cliff/issues/105 but, I can't build `cocogitto`, https://github.com/cocogitto/cocogitto using the same _"fix"_

@ethomson hopefully, the clean-up will propagate soon to `libgit2-sys` crate and make things reproducible.
Thanks!

**Comment by ethomson** [comment]
> @ethomson - I shouldn't have been so harsh - I know you are all doing a lot of work behind the scenes.

No need to apologize. I've been on the other side of the equation -- C library dependencies are annoying for C code and can be hell for everything else. This is especially true when the functionality is not core but very ancillary.

Thanks for the response.

I'll release a 1.5.2 soon - I've got a few other bugfixes that need to land first. 😓

**Comment by sstadick** [comment]
@0323pin - v0.9.0 is out now

**Comment by 0323pin** [comment]
Will kick start a build in 30 min or so, posting back the results.

**Comment by 0323pin** [comment]
All good again, thanks!

```
~ > hck -V
hck git:1d17af16e6
~ > pkgin list | hck -f1 > Git/netbsd/pkglist
~ > macchina

  Host       pin@mybox
  Machine    Hewlett-Packard HP EliteBook 820 G1 A3009DD10303
  OS         NetBSD 9.99.108
  WM         LeftWM
  Packages   191 (pkgsrc)
  Terminal   alacritty
  Shell      elvish
  Uptime     14m
  CPU        Intel® Core™ i5-4210U CPU @ 1.70GHz (4)
  Memory     2.5 GB/9.6 GB
  Battery    Full

   ●  ●  ●  ●  ●  ●  ●  ●

~ >                                                                                                                                                                                   < 19:55
```

**Comment by sstadick** [comment]
@0323pin - something is off, my publish CI failed, which I've fixed and it's trying again. But that version number you have there is from an old commit (the last tagged release). Are you cloning directly or pulling the archives from the release page?

Substantively, what you have is fine and equivalent to the commit for 0.9.0, it's showing the older commit just based on how it tracks back to a tag, but depending on how you are building this if you wanted to release one more time with v0.9.1 you should get a nicer version string. 

tl;dr; try with v0.9.1 again and you should get the following version string `hck git:v0.9.1` (lack of sha indicates it's on the same commit as the v0.9.1 tag)

**Comment by 0323pin** [comment]
> Are you cloning directly or pulling the archives from the release page?

I'm pulling from _tags_, since releases are stuck.

I'm building 0.9.1 right now, nearly there :)

**Comment by 0323pin** [comment]
@sstadick 0.9.1 merged into main now, http://mail-index.netbsd.org/pkgsrc-changes/2023/01/26/msg268190.html

I still get the hash,

![2023-01-26-204018_1366x768_scrot](https://user-images.githubusercontent.com/90570748/214935340-621850ec-f456-4129-8e82-2173af879ff7.png)

**Comment by sstadick** [comment]
Releases should be up to date now - good point on changelog!

Still weird. I don't think it's worth fussing over since it's still descriptive enough that we can track back to exact code running for that binary. But that's not very satisfying.

**Comment by 0323pin** [comment]
> Releases should be up to date now

Ops! I should check tomorrow if the checksums need to be updated, thanks for the heads-up.

> . I don't think it's worth fussing over since it's still descriptive enough that we can track back to exact code running for that binary. But that's not very satisfying.

True but, I can always get a diff between releases. I'd say, fix when time allows but, not  a priority. Thank you for the build fix.
